### ISSUE TITLE
Dynamo fails on bit shifts when dynamic=True

### ISSUE THREAD

**Issue by ar0ck** [pytorch/pytorch#97096]
### 🐛 Describe the bug

```
import torch

@torch.compile(dynamic=True, backend='eager')
def f(x):
    return x >> 0

f(torch.tensor(0))
```

The operations `>>` and  `<<` fail, while `{+, -, *, /, &, |, ^}` are OK.

### Error logs

```
Traceback (most recent call last):
  File "/.../.venv/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 1240, in dispatch
    r = func(*args, **kwargs)
  File "/.../.venv/lib/python3.9/site-packages/torch/_ops.py", line 330, in __call__
    return self._op(*args, **kwargs or {})
  File "/.../.venv/lib/python3.9/site-packages/torch/_ops.py", line 406, in _get_dispatch
    final_key = resolve_key(self, key)
  File "/.../.venv/lib/python3.9/site-packages/torch/_ops.py", line 186, in resolve_key
    raise NotImplementedError(f"could not find kernel for {op} at dispatch key {k}")
NotImplementedError: could not find kernel for aten.__rshift__.Scalar at dispatch key DispatchKey.Meta

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/utils.py", line 1195, in run_node
    return node.target(*args, **kwargs)
  File "/.../.venv/lib/python3.9/site-packages/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/.../.venv/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 1057, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "/.../.venv/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 1245, in dispatch
    return run_fallback_kernel(self, func, args, kwargs, not_implemented_error)
  File "/.../.venv/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 1430, in run_fallback_kernel
    args = tree_map(to_real_tensor, args)
  File "/.../.venv/lib/python3.9/site-packages/torch/utils/_pytree.py", line 196, in tree_map
    return tree_unflatten([fn(i) for i in flat_args], spec)
  File "/.../.venv/lib/python3.9/site-packages/torch/utils/_pytree.py", line 196, in <listcomp>
    return tree_unflatten([fn(i) for i in flat_args], spec)
  File "/.../.venv/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 1423, in to_real_tensor
    out = torch.zeros_like(e, device=e.fake_device)
RuntimeError: Cannot call strides() on tensor with symbolic sizes/strides

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/utils.py", line 1153, in get_fake_value
    return wrap_fake_exception(
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/utils.py", line 808, in wrap_fake_exception
    return fn()
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/utils.py", line 1154, in <lambda>
    lambda: run_node(tx.output, node, args, kwargs, nnmodule)
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/utils.py", line 1207, in run_node
    raise RuntimeError(
RuntimeError: Failed running call_function <built-in function rshift>(*(FakeTensor(FakeTensor(..., device='meta', size=(), dtype=torch.int64), cpu), 0), **{}):
Cannot call strides() on tensor with symbolic sizes/strides
(scroll up for backtrace)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/.../bug.py", line 7, in <module>
    f(torch.tensor(0))
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/eval_frame.py", line 235, in _fn
    return fn(*args, **kwargs)
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/eval_frame.py", line 372, in catch_errors
    return callback(frame, cache_size, hooks)
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/convert_frame.py", line 405, in _convert_frame
    result = inner_convert(frame, cache_size, hooks)
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/convert_frame.py", line 105, in _fn
    return fn(*args, **kwargs)
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/convert_frame.py", line 263, in _convert_frame_assert
    return _compile(
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/utils.py", line 164, in time_wrapper
    r = func(*args, **kwargs)
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/convert_frame.py", line 325, in _compile
    out_code = transform_code_object(code, transform)
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/bytecode_transformation.py", line 530, in transform_code_object
    transformations(instructions, code_options)
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/convert_frame.py", line 312, in transform
    tracer.run()
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 1848, in run
    super().run()
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 604, in run
    and self.step()
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 564, in step
    getattr(self, inst.opname)(inst)
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/symbolic_convert.py", line 151, in impl
    self.push(fn_var.call_function(self, self.popn(nargs), {}))
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/variables/builtin.py", line 515, in call_function
    return wrap_fx_proxy(tx, proxy, **options)
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/variables/builder.py", line 877, in wrap_fx_proxy
    return wrap_fx_proxy_cls(
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/variables/builder.py", line 912, in wrap_fx_proxy_cls
    example_value = get_fake_value(proxy.node, tx)
  File "/.../.venv/lib/python3.9/site-packages/torch/_dynamo/utils.py", line 1174, in get_fake_value
    raise TorchRuntimeError() from e
torch._dynamo.exc.TorchRuntimeError: 

from user code:
   File "/.../bug.py", line 5, in f
    return x >> 0


You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True
```

### Versions

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230318+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.2
Libc version: glibc-2.31

Python version: 3.9.5 (default, Nov 23 2021, 15:27:38)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.4.0-144-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: 10.1.243
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3090
Nvidia driver version: 530.30.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.8.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.8.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          32
On-line CPU(s) list:             0-31
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Silver 4314 CPU @ 2.40GHz
Stepping:                        6
Frequency boost:                 enabled
CPU MHz:                         1352.822
CPU max MHz:                     3400.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4800.00
Virtualization:                  VT-x
L1d cache:                       768 KiB
L1i cache:                       512 KiB
L2 cache:                        20 MiB
L3 cache:                        24 MiB
NUMA node0 CPU(s):               0-31
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid md_clear pconfig flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] torch==2.1.0.dev20230318+cpu
[conda] Could not collect
```

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by ezyang** [comment]
This is a coverage error, shouldn't be hard to fix. https://docs.google.com/document/d/1GgvOe7C8_NVOMLOCwDaYV1mXXyHMXY7ExoewHqooxrs/edit# says how to do it.

**Comment by shourya5** [comment]
I'll take this

**Comment by shourya5** [comment]
Just to clarify, the doc lists several approaches to solve this, According to me it should be to SymInt’ify C++ operator. Is that correct?

**Comment by ezyang** [comment]
Actually, the *first* error message is "NotImplementedError: could not find kernel for aten.__rshift__.Scalar at dispatch key DispatchKey.Meta", so I would do Python meta. 

Fortunately, there are already decomps in PrimTorch:

```
# TODO: add docstring
@_make_elementwise_binary_reference(
    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
)   
def bitwise_right_shift(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
    return prims.shift_right_arithmetic(a, b)
    
    
# TODO: add docstring
@_make_elementwise_binary_reference(
    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
)   
def bitwise_xor(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
    return prims.bitwise_xor(a, b)
```

So you just have to `@register_decomposition(aten.add)` these. I'll update the manual.

**Comment by ngimel** [comment]
Should we go through decompositions, or register unary pw meta directly in _meta_registrations? There's already a lot of funcs in `_meta_registrations` that return `empty_like(self)`, that should be enough for `__rshift__.Scalar`.   `__rshift__.Tensor` requires type promotion unfortunately.

**Comment by ezyang** [comment]
If you're concerned about perf, these operations should be doable via the binary fast path in fake tensor. For something quick and easy using the decomp seems fine though.

**Comment by ar0ck** [comment]
Fixed by #103637.
